### DISCUSSION TITLE
Issue with nullable Enum in generic method

### DISCUSSION THREAD

**Discussion by miroslav22** [dotnet/csharplang#6210] [General] [answered]
There appears to be a bug in the compiler when passing a nullable enum type with contraint of Enum:

`
namespace Test
{

    public class Program
    {
        [Flags]
        public enum FlagsEnum
        {
            Flag1 = 1 << 0,
            Flag2 = 1 << 1,
            Flag3 = 1 << 2,
        }

        static bool GetFlag<TEnum>(TEnum? field, TEnum flag) where TEnum : Enum => field?.HasFlag(flag) == true;

        public static void Main()
        {
            FlagsEnum? nullable = null;

            var result = GetFlag(nullable, FlagsEnum.Flag1);
        }
    }
}
`
The line:

var result = GetFlag(nullable, FlagsEnum.Flag1);

CS0312	The type 'Test.FlagsEnum?' cannot be used as type parameter 'TEnum' in the generic type or method 'Program.GetFlag<TEnum>(TEnum?, TEnum)'. The nullable type 'Test.FlagsEnum?' does not satisfy the constraint of 'System.Enum'.

Seems to be an incorrect compile error.

NET 6.0 / VS 17.2.0

**Comment by ufcpp** [comment]
`where TEnum : struct, Enum`

**Comment by PathogenDavid** [comment, marked as answer]
You need to constrain `TEnum` to `struct`. ([SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWAChKYIxUBnABwgGNUACAFQYUoG9KlAgGZ2xMSQDs/SuznsA2gDEANhADm9ALqz5I9qhgBXMO1Ub6AUWNhdcgRXlOza9SXYBedu4A8P9gAMcHbO5upEnt7sft7Bjs5yYaJevv5EcU4AvoLxcgQkAGzswAD2JSrsAOKoCGE+nNYmAHwAFA02APzsWFCoKgAmcFyNpliuAJQhTgDuABYYHO0m7CDs9AjoRiwIQyNT8h5N3b0DHQB0ABIQ9GEtYxrjnl4bRqgA3DlOxEQhDgmJrisnXYxhUamAKg4XlBKg+VFyzmQEHQ7HQDCMKgQkWqtVcLRhEAhqCGYSBJjOYRI4zhWUo2QomSAA===))

```csharp
static bool GetFlag<TEnum>(TEnum? field, TEnum flag)
    where TEnum : struct, Enum
    => field?.HasFlag(flag) == true;
```

**Reply by miroslav22** [reply]
Ah that's great, thanks.  Why is this out of interest?  Just a quirk of the compiler?

**Reply by PathogenDavid** [reply]
This is a quirk of how nullable reference types interact with generics. (You get a hint in the form of [the warning you get when nullable reference types are disabled](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAYhgFcAbUiYU1AAgBMoBnS6gWACgABAJgEYOOnAMw1OvAGyjuNACqpGCDgG8ONNaJFjJwAPY7SNAOKoEAMQoBzAMoJ0xAMYIAPDICiJMAD4AFG48B+GiwoVFI6OFl3YjAgywBKVXUkgHcACwxaP2iaEBoosESktQBeTyCQsP8AOgAJCEZzCAtvLHiaYuKaW2JUAG4OAF8OIA==).)

Without the `struct` constraint, `System.Enum` is valid for `TEnum`, and `System.Enum` is a reference type.

When a generic type can be either a reference type or a value type, the `?` only applies when it's a reference type.

Nullable reference types don't affect the type system. IE: As far as the runtime cares `object` and `object?` are the same thing.

Internally a nullable value type is just syntactic sugar for `System.Nullable<T>`. IE: `int?` is a different type entirely from `int`. So in a generic context, when `TEnum` can only be a struct, `TEnum?` actually compiles down to `System.Nullable<TEnum>`. ([You can see this here](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTCDMACBFANgKTwBUBTAZwBdMBvTPFg/Qk4Aey4Bs8A4pVoAxXhADmAZVoAnAK4BjWgB5yAURjywAPgAUGrWAD8eAGZRKvACYAaCpu3nxEgJTNWngO4ALSrMoHIzwQPDoFZXtHMA9PFgBeHXNLG2MAOgAJCGoxST0zF1c8ePi8OXlKAG5MAF9MIA=) -- compare to [this version without the constraint](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABAJgEYyzyBmXc6gNmctwBVsBnAFzIBvMrjHMmLdsAD2M1LgDi2fgDEMAcwDK/AE7IAxvwA8XAKIowAPgAU5ywH5cAMyjZUAEwA03C8jAumgCUouJhAO4AFti6ePb+uCC4fmChYWIAvFYubp4OAHQAEhC86hAaNs7BuBkZuHrI2ADcZAC+ZEA==)).

Because of these differences, there isn't a way for the compiler to represent a method that takes a nullable reference `T` *or* a nullable value `T`. It has to be one or the other.

**Comment by jnm2** [comment]
Same issue as in https://github.com/dotnet/csharplang/discussions/6004.
